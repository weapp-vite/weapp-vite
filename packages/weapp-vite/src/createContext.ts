import type { LoadConfigOptions } from './context'
import { getCompilerContext, resetCompilerContext } from './context/getInstance'

export async function createCompilerContext(options?: Partial<LoadConfigOptions & { key?: string }>) {
  // 先初始化 ConfigService
  const key = options?.key ?? 'default'
  if (!options?.key) {
    // ensure callers without explicit key do not reuse stale global context
    resetCompilerContext(key)
  }
  const ctx = getCompilerContext(key)
  const { configService, scanService } = ctx
  await configService.load(options)
  // prefilght
  try {
    await scanService.loadAppEntry()
  }
  catch {
    // prefilght catch
  }

  return ctx
}
