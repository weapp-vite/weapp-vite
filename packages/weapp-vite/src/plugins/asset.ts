import type { Buffer } from 'node:buffer'
import type { Plugin, ResolvedConfig } from 'vite'
import type { CompilerContext } from '../context'
import type { CopyGlobs } from '../types'
import { fdir as Fdir } from 'fdir'
import fs from 'fs-extra'
import path from 'pathe'
import { defaultAssetExtensions, defaultExcluded } from '../defaults'

interface AssetCandidate {
  file: string
  buffer: Buffer
}

interface AssetPluginState {
  ctx: CompilerContext
  resolvedConfig?: ResolvedConfig
  pendingAssets?: Promise<AssetCandidate[]>
}

export function asset(ctx: CompilerContext): Plugin[] {
  const state: AssetPluginState = { ctx }
  return [createAssetCollector(state)]
}

function createAssetCollector(state: AssetPluginState): Plugin {
  const { ctx } = state
  const { configService } = ctx

  return {
    name: 'weapp-vite:asset',
    enforce: 'pre',

    configResolved(config) {
      state.resolvedConfig = config
    },

    buildStart() {
      if (!state.resolvedConfig) {
        state.pendingAssets = Promise.resolve([])
        return
      }

      state.pendingAssets = scanAssetFiles(configService, state.resolvedConfig)
    },

    async buildEnd() {
      const assets = await state.pendingAssets
      if (!assets?.length) {
        return
      }

      for (const candidate of assets) {
        this.emitFile({
          type: 'asset',
          fileName: configService.relativeAbsoluteSrcRoot(candidate.file),
          source: candidate.buffer,
        })
      }
    },
  }
}

function scanAssetFiles(configService: CompilerContext['configService'], config: ResolvedConfig) {
  const weappViteConfig = configService.weappViteConfig
  const include = normalizeCopyGlobs(weappViteConfig?.copy?.include)
  const exclude = normalizeCopyGlobs(weappViteConfig?.copy?.exclude)
  const filter = weappViteConfig?.copy?.filter ?? (() => true)

  const ignore = [
    ...defaultExcluded,
    path.resolve(configService.cwd, `${config.build.outDir}/**/*`),
    ...exclude,
  ]

  const fdir = new Fdir({
    includeDirs: false,
    pathSeparator: '/',
  })

  const patterns = [
    `**/*.{${defaultAssetExtensions.join(',')}}`,
    ...include,
  ]

  return fdir
    .withFullPaths()
    .globWithOptions(patterns, {
      ignore,
      dot: false,
    })
    .crawl(configService.absoluteSrcRoot)
    .withPromise()
    .then((files) => {
      return Promise.all(
        files
          .filter(filter)
          .map(async (file) => {
            return {
              file,
              buffer: await fs.readFile(file),
            }
          }),
      )
    })
}

function normalizeCopyGlobs(globs?: CopyGlobs): string[] {
  return Array.isArray(globs) ? globs : []
}
