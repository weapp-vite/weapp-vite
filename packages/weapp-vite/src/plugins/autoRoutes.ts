import type { Plugin, ResolvedConfig } from 'vite'
import type { CompilerContext } from '../context'

const AUTO_ROUTES_ID = 'weapp-vite/auto-routes'
const VIRTUAL_MODULE_ID = 'virtual:weapp-vite-auto-routes'
const RESOLVED_VIRTUAL_ID = '\0weapp-vite:auto-routes'

function createAutoRoutesPlugin(ctx: CompilerContext): Plugin {
  const service = ctx.autoRoutesService
  let resolvedConfig: ResolvedConfig | undefined

  return {
    name: 'weapp-vite:auto-routes',
    enforce: 'pre',

    configResolved(config) {
      resolvedConfig = config
    },

    async buildStart() {
      await service.ensureFresh()
    },

    resolveId(id) {
      if (id === AUTO_ROUTES_ID || id === VIRTUAL_MODULE_ID) {
        return RESOLVED_VIRTUAL_ID
      }
      if (id === RESOLVED_VIRTUAL_ID) {
        return RESOLVED_VIRTUAL_ID
      }
      return null
    },

    async load(id) {
      if (id !== RESOLVED_VIRTUAL_ID) {
        return null
      }

      await service.ensureFresh()

      for (const file of service.getWatchFiles()) {
        try {
          this.addWatchFile(file)
        }
        catch { }
      }

      for (const dir of service.getWatchDirectories()) {
        try {
          this.addWatchFile(dir)
        }
        catch { }
      }

      return {
        code: service.getModuleCode(),
        map: { mappings: '' },
      }
    },

    async watchChange(id) {
      if (!service.isRouteFile(id)) {
        return
      }
      await service.handleFileChange(id)
    },

    async handleHotUpdate(context) {
      if (!service.isRouteFile(context.file)) {
        return
      }

      if (resolvedConfig?.command === 'serve') {
        await service.ensureFresh()
      }

      const virtualModule = context.server.moduleGraph.getModuleById(RESOLVED_VIRTUAL_ID)
      if (virtualModule) {
        return [virtualModule]
      }

      return context.modules.filter(module => module.id === RESOLVED_VIRTUAL_ID)
    },
  }
}

export function autoRoutes(ctx: CompilerContext): Plugin[] {
  return [createAutoRoutesPlugin(ctx)]
}
