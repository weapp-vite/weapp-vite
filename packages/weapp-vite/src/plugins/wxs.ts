import type { EmittedFile, PluginContext } from 'rolldown'
import type { Plugin } from 'vite'
import type { CompilerContext } from '../context'
import type { WxmlDep } from '../types'
import { removeExtension } from '@weapp-core/shared'
import fs from 'fs-extra'
import { LRUCache } from 'lru-cache'
import path from 'pathe'
import { jsExtensions } from '../constants'
import { transformWxsCode } from '../wxs'

export const wxsCodeCache = new LRUCache<string, string>(
  {
    max: 512,
  },
)

export function wxs({ configService, wxmlService }: CompilerContext): Plugin[] {
  const wxsMap = new Map<string, {
    emittedFile: EmittedFile
  }>()

  async function transformWxs(this: PluginContext, {
    wxsPath,
  }: {
    wxsPath: string
  }) {
    this.addWatchFile(wxsPath)
    if (await fs.exists(wxsPath)) {
      const arr = wxsPath.match(/\.wxs(\.[jt]s)?$/)
      let isRaw = true
      if (arr) {
        isRaw = !arr[1]
      }
      const rawCode = await fs.readFile(wxsPath, 'utf8')
      let code = wxsCodeCache.get(rawCode)

      const dirname = path.dirname(wxsPath)
      if (code === undefined) {
        const { result, importees } = transformWxsCode(rawCode, {
          filename: wxsPath,
        })
        if (typeof result?.code === 'string') {
          code = result.code
        }
        await Promise.all(
          importees.map(({ source }) => {
            return transformWxs.call(this, {
              wxsPath: path.resolve(dirname, source),
            })
          }),
        )
      }

      if (code !== undefined) {
        wxsMap.set(wxsPath, {
          emittedFile: {
            type: 'asset',
            fileName: configService.relativeAbsoluteSrcRoot(isRaw ? wxsPath : removeExtension(wxsPath)),
            source: code,
          },
        })
        wxsCodeCache.set(rawCode, code)
      }
    }
  }
  function handleWxsDeps(this: PluginContext, deps: WxmlDep[], absPath: string) {
    return Promise.all(
      deps.filter(x => x.tagName === 'wxs').map(async (wxsDep) => {
        const arr = wxsDep.value.match(/\.wxs(\.[jt]s)?$/)
        if (jsExtensions.includes(wxsDep.attrs.lang) || arr) {
          const wxsPath = path.resolve(path.dirname(absPath), wxsDep.value)

          await transformWxs.call(this, {
            wxsPath,
          })
        }
      }),
    )
  }

  return [
    {
      name: 'weapp-vite:wxs',
      enforce: 'pre',
      buildStart() {
        wxsMap.clear()
      },

      async buildEnd() {
        await Promise.all(
          Array.from(
            wxmlService
              .tokenMap
              .entries(),
          )
            .map(([id, token]) => {
              return handleWxsDeps.call(
                // @ts-ignore
                this,
                token.deps,
                id,
              )
            }),
        )

        Array.from(wxsMap.values()).forEach(({ emittedFile }) => {
          this.emitFile(emittedFile)
        })
      },
    },
  ]
}
