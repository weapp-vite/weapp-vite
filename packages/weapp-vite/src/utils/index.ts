export * from './file'
export * from './json'
export * from './md5'
export * from './projectConfig'
export * from './regexp'
export * from './version'
