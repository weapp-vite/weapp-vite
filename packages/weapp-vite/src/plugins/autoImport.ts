import type { Plugin, ResolvedConfig } from 'vite'
import type { CompilerContext } from '@/context'
import { fdir as Fdir } from 'fdir'
import path from 'pathe'
import { defaultExcluded } from '../defaults'
import { isTemplateRequest } from '../utils'

interface AutoImportState {
  ctx: CompilerContext
  resolvedConfig?: ResolvedConfig
  initialScanDone?: boolean
  lastGlobsKey?: string
}

export function autoImport(ctx: CompilerContext): Plugin[] {
  const state: AutoImportState = { ctx }
  return [createAutoImportPlugin(state)]
}

function createAutoImportPlugin(state: AutoImportState): Plugin {
  const { ctx } = state
  const { configService, autoImportService } = ctx

  return {
    name: 'weapp-vite:auto-import',
    enforce: 'pre',

    configResolved(config) {
      state.resolvedConfig = config
    },

    async buildStart() {
      if (!state.resolvedConfig) {
        return
      }

      const globs = configService.weappViteConfig?.enhance?.autoImportComponents?.globs
      const globsKey = globs?.join('\0') ?? ''
      if (globsKey !== state.lastGlobsKey) {
        state.initialScanDone = false
        state.lastGlobsKey = globsKey
      }

      if (!globs?.length) {
        return
      }

      if (state.initialScanDone) {
        return
      }

      autoImportService.reset()

      const files = await findTemplateCandidates(state, globs)
      await Promise.all(files.map(file => autoImportService.registerPotentialComponent(file)))

      state.initialScanDone = true
    },

    async watchChange(id, change) {
      if (!state.initialScanDone || !state.resolvedConfig) {
        return
      }

      const filePath = normalizeChangedPath(id)
      if (!filePath) {
        return
      }

      const absolutePath = resolveAbsolutePath(ctx, filePath)
      if (!absolutePath) {
        return
      }

      if (!matchesAutoImportGlobs(ctx, absolutePath)) {
        return
      }

      if (change.event === 'delete') {
        autoImportService.removePotentialComponent(absolutePath)
        return
      }

      await autoImportService.registerPotentialComponent(absolutePath)
    },
  }
}

async function findTemplateCandidates(state: AutoImportState, globs: string[]) {
  const { ctx, resolvedConfig } = state
  const { configService } = ctx

  if (!resolvedConfig) {
    return []
  }

  const ignore = [
    ...defaultExcluded,
    `${resolvedConfig.build.outDir}/**`,
  ]

  const fdir = new Fdir({
    includeDirs: false,
    filters: [
      (candidate: string) => {
        return isTemplateRequest(candidate)
      },
    ],
    pathSeparator: '/',
  })

  return await fdir
    .withFullPaths()
    .globWithOptions(globs.map(pattern => path.resolve(configService.absoluteSrcRoot, pattern)), {
      ignore,
    })
    .crawl(configService.absoluteSrcRoot)
    .withPromise()
}

function normalizeChangedPath(id: string) {
  if (!id || id.startsWith('\0')) {
    return undefined
  }

  const [pathWithoutQuery] = id.split('?')
  return pathWithoutQuery
}

function resolveAbsolutePath(ctx: AutoImportState['ctx'], candidate: string) {
  const { configService } = ctx
  if (!configService) {
    return undefined
  }

  if (path.isAbsolute(candidate)) {
    return candidate
  }

  const resolvedFromSrc = path.resolve(configService.absoluteSrcRoot, candidate)
  if (resolvedFromSrc.startsWith(configService.absoluteSrcRoot)) {
    return resolvedFromSrc
  }

  return path.resolve(configService.cwd, candidate)
}

function matchesAutoImportGlobs(ctx: AutoImportState['ctx'], candidate: string) {
  const { autoImportService, configService } = ctx
  if (!autoImportService || !configService) {
    return false
  }

  const targets = new Set<string>([
    candidate,
    configService.relativeCwd(candidate),
    configService.relativeAbsoluteSrcRoot(candidate),
  ])

  targets.add(`/${configService.relativeAbsoluteSrcRoot(candidate)}`)
  targets.add(candidate.replaceAll('\\', '/'))

  for (const target of targets) {
    if (!target) {
      continue
    }
    if (autoImportService.filter(target)) {
      return true
    }
  }

  return false
}
