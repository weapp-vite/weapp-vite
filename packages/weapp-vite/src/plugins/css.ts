import type { OutputAsset, OutputBundle } from 'rolldown'
import type { Plugin, ResolvedConfig } from 'vite'
import type { CompilerContext } from '../context'
import type { SubPackageStyleEntry } from '../types'
import fs from 'fs-extra'
import path from 'pathe'
import { changeFileExtension, isJsOrTs } from '../utils'
import { cssCodeCache, processCssWithCache, renderSharedStyleEntry } from './css/shared/preprocessor'
import { collectSharedStyleEntries, injectSharedStyleImports, toPosixPath } from './css/shared/sharedStyles'

export { cssCodeCache }

async function handleBundleEntry(
  this: any,
  bundle: OutputBundle,
  bundleKey: string,
  asset: OutputAsset | OutputBundle[string],
  configService: CompilerContext['configService'],
  sharedStyles: Map<string, SubPackageStyleEntry[]>,
  emitted: Set<string>,
) {
  if (asset.type !== 'asset') {
    return
  }

  const toAbsolute = (id: string) => {
    return path.isAbsolute(id) ? id : path.resolve(configService.cwd, id)
  }

  if (bundleKey.endsWith('.wxss')) {
    const [rawOriginal] = asset.originalFileNames ?? []
    const absOriginal = rawOriginal
      ? toAbsolute(rawOriginal)
      : path.resolve(configService.absoluteSrcRoot, bundleKey)
    const fileName = configService.relativeAbsoluteSrcRoot(absOriginal)

    if (fileName) {
      emitted.add(toPosixPath(fileName))
    }

    if (fileName && fileName !== bundleKey) {
      delete bundle[bundleKey]
      const css = await fs.readFile(absOriginal, 'utf8')
      this.emitFile({
        type: 'asset',
        fileName,
        source: css,
      })
    }

    return
  }

  if (!bundleKey.endsWith('.css')) {
    return
  }

  if (!asset.originalFileNames) {
    delete bundle[bundleKey]
    return
  }

  await Promise.all(
    asset.originalFileNames.map(async (originalFileName) => {
      if (!isJsOrTs(originalFileName)) {
        return
      }

      const modulePath = toAbsolute(originalFileName)
      const converted = changeFileExtension(modulePath, configService.outputExtensions.wxss)
      const fileName = configService.relativeAbsoluteSrcRoot(converted)
      if (!fileName) {
        return
      }
      const normalizedFileName = toPosixPath(fileName)
      const rawCss = asset.source.toString()
      const processedCss = await processCssWithCache(rawCss, configService)

      const cssWithImports = injectSharedStyleImports(
        processedCss,
        modulePath,
        fileName,
        sharedStyles,
        configService,
      )

      this.emitFile({
        type: 'asset',
        fileName,
        source: cssWithImports,
      })
      emitted.add(normalizedFileName)
    }),
  )

  delete bundle[bundleKey]
}

async function emitSharedStyleEntries(
  this: any,
  sharedStyles: Map<string, SubPackageStyleEntry[]>,
  emitted: Set<string>,
  configService: CompilerContext['configService'],
  bundle: OutputBundle,
  resolvedConfig?: ResolvedConfig,
) {
  if (!sharedStyles.size) {
    return
  }

  for (const entries of sharedStyles.values()) {
    for (const entry of entries) {
      const fileName = toPosixPath(entry.outputRelativePath)
      if (emitted.has(fileName)) {
        continue
      }

      const absolutePath = entry.absolutePath
      if (typeof this.addWatchFile === 'function') {
        this.addWatchFile(absolutePath)
      }

      if (!await fs.pathExists(absolutePath)) {
        continue
      }

      const { css: renderedCss, dependencies } = await renderSharedStyleEntry(entry, configService, resolvedConfig)
      if (typeof this.addWatchFile === 'function' && dependencies.length) {
        for (const dependency of dependencies) {
          if (dependency && dependency !== absolutePath) {
            this.addWatchFile(dependency)
          }
        }
      }

      const css = await processCssWithCache(renderedCss, configService)

      emitted.add(fileName)
      if (bundle[fileName]) {
        delete bundle[fileName]
      }

      this.emitFile({
        type: 'asset',
        fileName,
        source: css,
      })
    }
  }
}

async function emitSharedStyleImportsForChunks(
  this: any,
  sharedStyles: Map<string, SubPackageStyleEntry[]>,
  emitted: Set<string>,
  configService: CompilerContext['configService'],
  bundle: OutputBundle,
) {
  if (!sharedStyles.size) {
    return
  }

  const { outputExtensions } = configService

  await Promise.all(
    Object.values(bundle).map(async (output) => {
      if (output.type !== 'chunk') {
        return
      }

      const moduleId = output.facadeModuleId
      if (!moduleId) {
        return
      }

      const relativeModulePath = configService.relativeAbsoluteSrcRoot(moduleId)
      if (!relativeModulePath) {
        return
      }

      const converted = changeFileExtension(moduleId, outputExtensions.wxss)
      const fileName = configService.relativeAbsoluteSrcRoot(converted)
      if (!fileName) {
        return
      }

      const normalizedFileName = toPosixPath(fileName)
      if (emitted.has(normalizedFileName)) {
        return
      }

      const cssWithImports = injectSharedStyleImports(
        '',
        moduleId,
        fileName,
        sharedStyles,
        configService,
      )

      if (!cssWithImports.trim()) {
        return
      }

      const processedCss = await processCssWithCache(cssWithImports, configService)

      this.emitFile({
        type: 'asset',
        fileName,
        source: processedCss,
      })

      emitted.add(normalizedFileName)
    }),
  )
}

async function generateBundleSharedCss(
  this: any,
  ctx: CompilerContext,
  configService: CompilerContext['configService'],
  bundle: OutputBundle,
  resolvedConfig?: ResolvedConfig,
) {
  const sharedStyles = collectSharedStyleEntries(ctx, configService)
  const emitted = new Set<string>()
  const tasks = Object.entries(bundle).map(([bundleKey, asset]) => {
    return handleBundleEntry.call(this, bundle, bundleKey, asset, configService, sharedStyles, emitted)
  })

  await Promise.all(tasks)
  await emitSharedStyleEntries.call(this, sharedStyles, emitted, configService, bundle, resolvedConfig)
  await emitSharedStyleImportsForChunks.call(this, sharedStyles, emitted, configService, bundle)
}

export function css(ctx: CompilerContext): Plugin[] {
  const { configService } = ctx
  let resolvedConfig: ResolvedConfig | undefined
  return [
    {
      name: 'weapp-vite:css',
      enforce: 'pre',
      configResolved(config) {
        resolvedConfig = config
      },
      async generateBundle(_opts, bundle) {
        await generateBundleSharedCss.call(this, ctx, configService, bundle, resolvedConfig)
      },
    },
  ]
}
