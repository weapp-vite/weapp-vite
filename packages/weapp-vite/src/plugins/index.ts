import type { Plugin } from 'vite'
import type { CompilerContext } from '@/context'
import type { SubPackageMetaValue, WeappVitePluginApi } from '@/types'
import { wrapPlugin } from 'vite-plugin-performance'
import { asset } from './asset'
import { autoImport } from './autoImport'
import { autoRoutes } from './autoRoutes'
import { weappVite } from './core'
import { css } from './css'
import { preflight } from './preflight'
import { workers } from './workers'
import { wxs } from './wxs'

const RUNTIME_PLUGINS_SYMBOL = Symbol.for('weapp-runtime:plugins')

function attachRuntimePlugins(ctx: CompilerContext, plugins: Plugin[]): Plugin[] {
  const runtimePlugins = (ctx as any)[RUNTIME_PLUGINS_SYMBOL] as Plugin[] | undefined
  if (!runtimePlugins?.length) {
    return plugins
  }

  return [...runtimePlugins, ...plugins]
}

function applyInspect(ctx: CompilerContext, plugins: Plugin[]): Plugin[] {
  const inspectOptions = ctx.configService.weappViteConfig?.debug?.inspect
  if (!inspectOptions) {
    return plugins
  }

  // @ts-ignore third-party typings do not surface Plugin[] overload
  return wrapPlugin(plugins, inspectOptions)
}

function flatten(groups: Plugin[][]): Plugin[] {
  return groups.reduce<Plugin[]>((acc, cur) => {
    acc.push(...cur)
    return acc
  }, [])
}

export function vitePluginWeapp(
  ctx: CompilerContext,
  subPackageMeta?: SubPackageMetaValue,
): Plugin<WeappVitePluginApi>[] {
  const groups: Plugin[][] = [preflight(ctx)]

  if (!subPackageMeta) {
    groups.push(asset(ctx), autoRoutes(ctx), autoImport(ctx))
  }

  groups.push(weappVite(ctx, subPackageMeta), wxs(ctx), css(ctx))

  const assembled = attachRuntimePlugins(ctx, flatten(groups))
  if (subPackageMeta) {
    return assembled
  }

  return applyInspect(ctx, assembled)
}

export function vitePluginWeappWorkers(ctx: CompilerContext) {
  const groups = [preflight(ctx), workers(ctx)]
  const assembled = attachRuntimePlugins(ctx, flatten(groups))
  return applyInspect(ctx, assembled)
}
