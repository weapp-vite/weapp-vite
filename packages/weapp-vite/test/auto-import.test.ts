import CI from 'ci-info'
import fs from 'fs-extra'
import path from 'pathe'
import { createCompilerContext } from '@/createContext'
import { cssCodeCache } from '@/plugins/css'
import { wxsCodeCache } from '@/plugins/wxs'
import { getFixture, scanFiles } from './utils'

describe.skipIf(CI.isCI)('auto-import', () => {
  const cwd = getFixture('auto-import')
  const distDir = path.resolve(cwd, 'dist')
  beforeAll(async () => {
    await fs.remove(distDir)
    const ctx = await createCompilerContext({
      cwd,
    })
    await ctx.buildService.build()
    expect(await fs.exists(distDir)).toBe(true)
  })

  it('scanFiles', async () => {
    const files = await scanFiles(distDir)
    expect(files).toMatchSnapshot()
    const codes = await Promise.all(
      files.filter(x => x.endsWith('.json')).map((x) => {
        return fs.readFile(path.resolve(distDir, x), 'utf-8')
      }),
    )
    expect(codes).toMatchSnapshot()
    expect(wxsCodeCache.size).toBe(0)
    expect(cssCodeCache.size).toBe(8)
  })
})
