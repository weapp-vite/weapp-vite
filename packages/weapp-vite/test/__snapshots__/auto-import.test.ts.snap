// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auto-import > scanFiles 1`] = `
[
  "app.js",
  "app.json",
  "app.wxss",
  "components/Avatar/Avatar.js",
  "components/Avatar/Avatar.json",
  "components/Avatar/Avatar.wxml",
  "components/Avatar/Avatar.wxss",
  "components/HelloWorld/index.js",
  "components/HelloWorld/index.json",
  "components/HelloWorld/index.wxml",
  "components/HelloWorld/index.wxss",
  "components/HiChina/HiChina.js",
  "components/HiChina/HiChina.json",
  "components/HiChina/HiChina.wxml",
  "components/HiChina/HiChina.wxss",
  "components/Navbar/Navbar.js",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.wxml",
  "components/Navbar/Navbar.wxss",
  "components/XxYy/XxYy.js",
  "components/XxYy/XxYy.json",
  "components/XxYy/XxYy.wxml",
  "components/XxYy/XxYy.wxss",
  "components/icebreaker/index.js",
  "components/icebreaker/index.json",
  "components/icebreaker/index.wxml",
  "components/icebreaker/index.wxss",
  "components/xx-yy/index.js",
  "components/xx-yy/index.json",
  "components/xx-yy/index.wxml",
  "components/xx-yy/index.wxss",
  "pages/index/index.js",
  "pages/index/index.json",
  "pages/index/index.wxml",
  "pages/index/index.wxss",
  "sitemap.json",
  "theme.json",
]
`;

exports[`auto-import > scanFiles 2`] = `
[
  "{
  "pages": [
    "pages/index/index",
    "pages/index/vue"
  ],
  "window": {},
  "style": "v2",
  "componentFramework": "glass-easel",
  "sitemapLocation": "sitemap.json"
}",
  "{
  "component": true,
  "usingComponents": {
    "van-grid-item": "@vant/weapp/grid-item",
    "van-grid": "@vant/weapp/grid"
  }
}",
  "{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}",
  "{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {
    "van-button": "@vant/weapp/button",
    "van-slider": "@vant/weapp/slider",
    "van-switch": "@vant/weapp/switch",
    "van-circle": "@vant/weapp/circle",
    "van-card": "@vant/weapp/card"
  }
}",
  "{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}",
  "{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}",
  "{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {
    "HiChina": "/components/HiChina/HiChina",
    "HelloWorld": "/components/HelloWorld/index"
  }
}",
  "{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}",
  "{
  "usingComponents": {
    "Avatar": "../../components/Avatar/Avatar",
    "Navbar": "/components/Navbar/Navbar",
    "HelloWorld": "/components/HelloWorld/index",
    "icebreaker": "/components/icebreaker/index",
    "xx-yy": "/components/xx-yy/index",
    "XxYy": "/components/XxYy/XxYy"
  },
  "navigationBarTitleText": "初始模板"
}",
  "{
  "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
  "rules": [
    {
      "action": "allow",
      "page": "*"
    }
  ]
}",
  "{
  "light": {},
  "dark": {}
}",
]
`;
